src/textual/widgets/_option_list.py

```diff
@@ -0,0 +1,904 @@
+
+    def _set_option_disabled(self, index: int, disabled: bool) -> Self:
+        """Set the disabled state of an option in the list.
+
+        Args:
+            index: The index of the option to set the disabled state of.
+            disabled: The disabled state to set.
+
+        Returns:
+            The `OptionList` instance.
+        """
+        self._options[index].disabled = disabled
+        # TODO: Refresh only if the affected option is visible.
+        self.refresh()
+        return self
+
+    def enable_option_at_index(self, index: int) -> Self:
+        """Enable the option at the given index.
+
+        Returns:
+            The `OptionList` instance.
+
+        Raises:
+            IndexError: If there is no option with the given index.
+        """
+        return self._set_option_disabled(index, False)
+
+    def disable_option_at_index(self, index: int) -> Self:
+        """Disable the option at the given index.
+
+        Returns:
+            The `OptionList` instance.
+
+        Raises:
+            IndexError: If there is no option with the given index.
+        """
+        return self._set_option_disabled(index, True)
+
+    def enable_option(self, option_id: str) -> Self:
+        """Enable the option with the given ID.
+
+        Args:
+            option_id: The ID of the option to enable.
+
+        Returns:
+            The `OptionList` instance.
+
+        Raises:
+            KeyError: If no option has the given ID.
+        """
+        return self.enable_option_at_index(self._option_ids[option_id])
+
+    def disable_option(self, option_id: str) -> Self:
+        """Disable the option with the given ID.
+
+        Args:
+            option_id: The ID of the option to disable.
+
+        Returns:
+            The `OptionList` instance.
+
+        Raises:
+            KeyError: If no option has the given ID.
+        """
+        return self.disable_option_at_index(self._option_ids[option_id])
+
+    @property
+    def option_count(self) -> int:
+        """The count of options."""
+        return len(self._options)
+
+    @property
+    def options(self) -> Iterator[Option]:
+        """An iterator of options within the list."""
+        return iter(self._options)
+
+    def get_option_at_index(self, index: int) -> Option:
+        """Get the option at the given index.
+
+        Args:
+            index: The index of the option to get.
+
+        Returns:
+            The option at that index.
+
+        Raises:
+            IndexError: If there is no option with the index.
+        """
+        return self._options[index]
+
+    def get_option(self, option_id: str) -> Option:
+        """Get the option with the given ID.
+
+        Args:
+            index: The ID of the option to get.
+
+        Returns:
+            The option at with the ID.
+
+        Raises:
+            KeyError: If no option has the given ID.
+        """
+        return self.get_option_at_index(self._option_ids[option_id])
```

Comment:
I think we might want to have a dedicated exception here. Probably the same exception as the previous method.

Reply:
Done.